core/src/test/java/org/elasticsearch/discovery/ZenFaultDetectionTests.java

```diff
@@ -192,6 +206,39 @@ public void onNodeFailure(DiscoveryNode node, String reason) {
         assertThat(failureReason[0], matcher);
     }
 
+    public void testMasterFaultDetectionNotSizeLimited() throws InterruptedException {
+        Settings.Builder settings = Settings.builder();
+        boolean shouldRetry = randomBoolean();
+        settings.put(FaultDetection.CONNECT_ON_NETWORK_DISCONNECT_SETTING.getKey(), shouldRetry)
+            .put(FaultDetection.PING_INTERVAL_SETTING.getKey(), "2s");
+        ClusterName clusterName = new ClusterName(randomAsciiOfLengthBetween(3, 20));
+        final ClusterState stateNodeA = ClusterState.builder(clusterName).nodes(buildNodesForA(false)).build();
+        setState(clusterServiceA, stateNodeA);
+        MasterFaultDetection masterFDNodeA = new MasterFaultDetection(settings.build(), threadPool, serviceA, clusterName,
+            clusterServiceA);
+        masterFDNodeA.start(nodeB, "test");
+
+        final ClusterState stateNodeB = ClusterState.builder(clusterName).nodes(buildNodesForB(true)).build();
+        setState(clusterServiceB, stateNodeB);
+
+        MasterFaultDetection masterFDNodeB = new MasterFaultDetection(settings.build(), threadPool, serviceB, clusterName,
+            clusterServiceB);
+        masterFDNodeB.start(nodeB, "test");
+
+        AtomicInteger failureCount = new AtomicInteger();
+        final CountDownLatch notified = new CountDownLatch(1);
+        masterFDNodeA.addListener((masterNode, cause, reason) -> {
+            failureCount.incrementAndGet();
+            notified.countDown();
+        });
+        // we have no callback when a ping has succeeded so we just let a few pings through
```

Comment:
But we should be able to hook into the transport to verify that a ping was received? I'd rather see a stronger guarantee here than sleeping for ten seconds?